wip on usage command

uovobw opened the pull request:
this code does _not_ compile at the moment.
My plan is to be able to create all the commands and the command dispatcher _outside_ of the loop in the main thread and use their reference as a command to be able to address them in the command dispatcher. i am not sure this is the right way to go, i fear it might require us to rework the whole command interface. any comments are welcome.

the change in itself is not huge, since rust does not support variadic functions and iterating on a Vec of Commands cannot be done (or i haven't been able to) the plan was to get the usage at the beginning of the main() and then return that each time it's needed. 
missing:
- privmsg handling
- destination type routing in both telegram and irc tranports
- decent logging

uovobw committed:
wip on usage command

crisidev commented:
I'll take a look to it later or maybe tomorrow. 
If I get it right the idea is just to reuse the commands and the dispatcher in the main loop without instantiating them every time.. I  think it is the right way to approach the problem. About the implementation, this is a playground and for me rewriting the full command interface is something we will have to do to allow mote reusable code!

crisidev commented on holysee/src/commands/command_dispatcher.rs:
```diff
@@ -7,30 +7,34 @@ use message::Message;
 
 impl<'a> CommandDispatcher<'a> {
-    pub fn new(settings: &'a settings::Commands) -> CommandDispatcher<'a> {
+    pub fn new(settings: &'a settings::Commands, base_command: &'a Command) -> CommandDispatcher<'a> {
```
Do we really need to pass the `base_command` to the constructor?

crisidev commented on holysee/src/commands/usage.rs:
```diff
@@ -0,0 +1,84 @@
+
+use self::regex::Regex;
+
+use message::{Message, TransportType,DestinationType};
```
Missing space between `TransportType` and `DestinationType`

crisidev reviewed (approved):
This looks good to me.. I like having commands initialisation inside outside of the loop.

uovobw committed:
fix creation/deletion from scope in main

uovobw committed:
typo

uovobw committed:
fmt

uovobw committed:
refactor the command executor, now all commands are instantiated once
and run many times

uovobw committed:
remove dead code

uovobw committed:
fix regex

uovobw committed:
some fmt
fix regexp formatting, it does not like regexp that overflow to the new
line
BUG: the messages and the output of all the commands is relayed to the
other transport no matter what

uovobw commented:
some more work in progress. abstracted the name and moved some functions in the `Command` trait.
there is a bug now introduced with the refactor: the if/else/if/else... chain in the main is gone, so now all commands are relayed no matted what, we need to find a way of marking the filter (like the last seen or the relay commands) so that it runs differently than the normal command

uovobw committed:
disable tests, rewrite them after the refactor

uovobw committed:
fix the relay problem with the stop_processing() Command trait function

crisidev commented:
LGTM. I think we can merge this!
